To evaluate an expression using Grover's Search Algorithm, we need to input the variables as letters. For example, in the problem below, we use the letters A, B, and C to represent different variables. Use & for AND. Use ~ for NOT. After running, you will get a list of numbers, each with different values. Each number represents a different solution to the problem, and the number with the highest values represent the solutions to the problem.

In [2]:
# Importing standard Qiskit libraries
from qiskit import QuantumCircuit, transpile
from qiskit.tools.jupyter import *
from qiskit.visualization import *
from ibm_quantum_widgets import *

from qiskit_ibm_runtime import QiskitRuntimeService, Sampler, Estimator, Session, Options

# Loading your IBM Quantum account(s)
service = QiskitRuntimeService(channel="ibm_quantum")

In [4]:
import numpy as np

from qiskit import *
from qiskit.circuit.library import PhaseOracle
from qiskit.algorithms import Grover, AmplificationProblem
from qiskit.primitives import Sampler



expression = '(A & B) & ~(B & C)' # change these variables to fit your problem
oracle = PhaseOracle(expression)
problem = AmplificationProblem(oracle=oracle, is_good_state=oracle.evaluate_bitstring)
backend = Aer.get_backend('qasm_simulator')
grover = Grover(sampler=Sampler())
result = grover.amplify(problem)
print(result.circuit_results)

[{'000': 0.0312499999999999, '001': 0.0312499999999999, '010': 0.0312499999999999, '011': 0.7812499999999987, '100': 0.0312499999999999, '101': 0.0312499999999999, '110': 0.0312499999999999, '111': 0.0312499999999999}]




The expression '(A & B) & ~(B & C)' gives us the following: {'000': 0.0312499999999999, '001': 0.0312499999999999, '010': 0.0312499999999999, '011': 0.7812499999999987, '100': 0.0312499999999999, '101': 0.0312499999999999, '110': 0.0312499999999999, '111': 0.0312499999999999} The number 011 has the highest value, 0.78, and thus is the solution to our expression. Note that the numbers start from the right so should interpreted backward. In this case, the number '011' represents the solution NOT C, B, and A.

To try your own expression please find the jupyter notebook at https://github.com/positivetechnologylab/groversearch